src/csync/csync_exclude.cpp

```diff
@@ -320,106 +322,249 @@ static CSYNC_EXCLUDE_TYPE _csync_excluded_common(c_strlist_t *excludes, const ch
         goto out;
     }
 
-    if (check_leading_dirs) {
-        /* Build a list of path components to check. */
-        path_components = c_strlist_new(32);
-        char *path_split = strdup(path);
-        size_t len = strlen(path_split);
-        for (i = len; ; --i) {
-            // read backwards until a path separator is found
-            if (i != 0 && path_split[i-1] != '/') {
-                continue;
-            }
+  out:
+    return match;
+}
 
-            // check 'basename', i.e. for "/foo/bar/fi" we'd check 'fi', 'bar', 'foo'
-            if (path_split[i] != 0) {
-                c_strlist_add_grow(&path_components, path_split + i);
-            }
+CSYNC_EXCLUDE_TYPE csync_excluded_traversal(CSYNC *ctx, const char *path, int filetype) {
+    CSYNC_EXCLUDE_TYPE match = CSYNC_NOT_EXCLUDED;
+    CSYNC_EXCLUDE_TYPE type  = CSYNC_NOT_EXCLUDED;
+    int rc = -1;
 
-            if (i == 0) {
-                break;
-            }
 
-            // check 'dirname', i.e. for "/foo/bar/fi" we'd check '/foo/bar', '/foo'
-            path_split[i-1] = '\0';
-            c_strlist_add_grow(&path_components, path_split);
-        }
-        SAFE_FREE(path_split);
+    match = _csync_excluded_common(ctx->excludes, path, filetype);
+    if (match != CSYNC_NOT_EXCLUDED) {
+        return match;
+    }
+    //qDebug() << path;
+
+    const char *bname = NULL;
+    size_t blen = 0;
+    /* split up the path */
+    bname = strrchr(path, '/');
+    if (bname) {
+        bname += 1; // don't include the /
+    } else {
+        bname = path;
     }
+    blen = strlen(bname);
+    QString p = QString::fromUtf8(bname);
 
+    if (ctx->parsed_traversal_excludes.regexp_exclude.match(p).hasMatch()) {
+        //qDebug() << "WOULD EXCLUDE";
+        return CSYNC_FILE_EXCLUDE_LIST;
+    }
+    if (ctx->parsed_traversal_excludes.regexp_exclude_and_remove.match(p).hasMatch()) {
+        //qDebug() << "WOULD EXCLUDE AND REMOVE";
+        return CSYNC_FILE_EXCLUDE_AND_REMOVE;
```

Comment:
could we use one regexp by using captures? `(exclude|these)|(exclude|and|remove|these)` and then check which capture is nonempty? Unclear whether that'd be faster